internal/inputter/inputter.go

```diff
@@ -91,27 +86,85 @@ func (w InputterWorker) watchNewInputs(
 	client *ethclient.Client,
 	inputBox *contracts.InputBox,
 ) error {
-	logs := make(chan *contracts.InputBoxInputAdded)
-	opts := bind.WatchOpts{
-		Context: ctx,
-	}
-	filter := []common.Address{w.ApplicationAddress}
-	sub, err := inputBox.WatchInputAdded(&opts, logs, filter, nil)
-	if err != nil {
-		return fmt.Errorf("inputter: watch input added: %w", err)
-	}
-	defer sub.Unsubscribe()
+	var sub ethereum.Subscription
+	var err error
+	seconds := 5
+	reconnectDelay := time.Duration(seconds) * time.Second
+	retryCount := 0
+	maxRetry := 3
+	currentBlock := w.InputBoxBlock
 	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case err := <-sub.Err():
-			return err
-		case event := <-logs:
-			if err := w.addInput(ctx, client, event); err != nil {
-				return err
+		// First, read the event logs to get the past inputs; then, watch the event logs to get the
+		// new ones. There is a race condition where we might lose inputs sent between the
+		// readPastInputs call and the watchNewInputs call. Given that nonodo is a development node,
+		// we accept this race condition.
+		err = w.readPastInputs(ctx, client, inputBox, currentBlock)
+		if err != nil {
+			if retryCount > maxRetry {
+				return fmt.Errorf("readPastInputs: %w", err)
+			} else {
+				slog.Warn("Inputter", "error", err)
+				slog.Info("Inputter reconnecting", "reconnectDelay", reconnectDelay)
```

Comment:
Why need this? I'd like know why this change in behaviour?

Reply:
https://github.com/Calindra/nonodo/issues/80